src/component/actionBar/index.jsx

```diff
@@ -1,19 +1,50 @@
-import * as React from 'react';
-import {useState} from "react";
-import {removeToken} from "../../utils/handleToken/handleToken";
+import React, {useEffect, useState} from 'react';
+import {removeToken} from "../../utils/handleToken";
 import {useHistory} from "react-router";
 import {NavLink} from "react-router-dom";
-import {defaultUserMenu} from "./serviceMenu";
+import {defaultUserMenu, GroupMasterMenu} from "./serviceMenu";
+import {fetchGroupMe} from "../../store/api/groupApi";
+import {fetchUserProfile} from "../../store/api/usersApi";
+import {jobLists} from "../../constants/jobLists";
 
 function ActionBar() {
   const history = useHistory();
-  const [userMenu, setUserMenu] = useState(defaultUserMenu);
+  const [userMenu, setUserMenu] = useState(GroupMasterMenu);
+  const [userName, setUserName] = useState('');
+  const [userGroupName, setUserGroupName] = useState('');
+  const [userDepartment, setUserDepartment] = useState('');
+  const [userJobName, setUserJobName] = useState('');
+  const [userRentals, setUserRentals] = useState(0);
+  const [userOverdueRentals, setUserOverdueRentals] = useState(0);
 
   function logout() {
     removeToken();
     history.replace('/register/signIn');
   }
 
+  useEffect(() => {
+    fetchUserProfile().then(data => {
+      const {name} = data;
+      setUserName(name);
+    });
+    fetchGroupMe().then(data => {
+      const {name: groupName, department, jobKey, rentals, overdueRentals} = data.results[0];
+      setUserGroupName(groupName);
+      setUserDepartment(department || '부서명');
+      setJobName(jobKey);
+      setUserRentals(rentals);
+      setUserOverdueRentals(overdueRentals);
+    })
+  },[userRentals, userOverdueRentals]);
```

Comment:
`useEffect` 내부에서 async await 를 하니까 경고를 해서 이렇게 하신거죠?

```
useEffect(() => {
  (async () => {
     const user = await fetchUserProfile();
     .. user 저장
     const group = await fetchGroupMe();
     ... data 저장
  }
}, []);
```

이렇게 async await 를 써도 되고, 아니면 fetchUserProfile과 then을 하나의 함수로 만드는 것도 좋아요.
(하지만 현재는 component state에 저장해야하기때문에 비추천하는 방법.) - 하나의 방법이라는걸로 이해하시면 될 것 같아요.

```
// utils/api.js
import * as API from 'api.js'

async function fetchUserProfile() {
   try {
      const user = API.fetchUserProfile();
      const { name } = data;
      setUserName(name);
   } catch(e) {
      // ...error처리
   }
} 

async function fetchGroupMe() { ... }

// actionBar/index.jsx

useEffect(() => {
  fetchUserProfile();
  fetchGroupMe();
}, []);
```

즉, API를 호출하는 함수를 만들라고 가이드를 드렸는데, API를 다루는 함수는 총 2개가 생기는거에요.
1. axios를 이용하여 API를 호출하는 함수.
2. 1번 함수를 사용하면서 성공, 실패처리를 하는 함수.

1번을 해야하는 이유는 API별로 특별히 맞추어야하는 헤더, 데이터 값이 특수한 경우에 헷갈리지 않게끔 하나의 함수로 만들어서 관리를 하는 의미.
1번에서 성공 실패처리를 하지 않고 2번 함수를 따로 만드는 이유는 API요청이 한번만 발생하지 않고 어느 곳에서든 요청해서 성공/실패 처리를 할 수 있기 때문에 1번과 2번이 분리되어서 각각 만들어져야해요.

> (하지만 현재는 component state에 저장해야하기때문에 비추천하는 방법.) - 하나의 방법이라는걸로 이해하시면 될 것 같아요.

여기서 할 수 있는 방법은 컴포넌트에서 호출하는 인자로 setState를 넘겨서 사용하면 되긴 해요. (할지 말지는 선택에 맡길게요.)

```
// utils/api.js
async function fetchUserProfile(success, failure) {
    try {
      const user = API.fetchUserProfile();
      const { name } = data;
      
      success(name);
    } catch(e) {
      failure && failure(e);
    }
} 


// actionBar/index.jsx

useEffect(() => {
   fetchUserProfile(handleSaveName);
   fetchGroupMe(handleSaveGroup);
}

function handleSaveName(name) { setUserName(name); }
function handleGroupName(data) {
   const { name, department, jobKey, .. } = data.results[0];

   setUserGroupName(groupName);
   ...
}
```

Reply:
Promise말고 async/await를 가이드 드리는 이유는

```
try {
const data = await fetchUserName();

setData(data);

const group = await fetchGroupMe();

setGroupData(group);
} catch(e) { ... }
```

```
fetchUserName()
.then((name) => {
  setName(name);
})
.then(fetchGroupMe)
.then((group) => {
  setGroupData(group);
})
.catch(e => { ... }) 
```

API 2개를 순차적으로 호출하려면 이렇게 두개로 나뉘는데, Promise의 경우는 return을 잘못하면 동작을 안하기도 하고 실수할 여지가 있어서 async/await 로 하는것이 낫다고 개인적으로 생각하고 있고, 그래서 Promise보단 async/await로 코드를 짜는것이 좋겠다고 가이드를 드리는 것입니다.

Reply:
```suggestion
  },[userRentals, userOverdueRentals]);
```
이렇게 작성한 이유는
제가 이해한 useEffect hook사용법에서는 두번째 인자인 배열이 빈 배열[]일 때는 최초의 한번만 랜더링하고, state나 props에 의존성이 없으면서 useEffect안에 사용되는 변수나 상수? 를 넣어야 인자들을 추적해 효율적으로 뷰를 바꾼다고 이해했기 때문입니다..! 최종적으로 actionBar 안에서 실시간으로 관찰해서 바꿔야하는 state값이 useRentals(대여중인 책 수)와 userOverdueRentals(연체된 책 수) 였기 때문에 저렇게 작성했습니다! 그래서 해당 api가 asnyc안에 들어간 후에는 한번만 부르고 끝나기 때문에 값이 바뀌었을때 뷰가 업데이트 되지 않지는 않을까요..?

Reply:
아, 제가 말씀드린 `이렇게 작성한`에 대한 내용은, `useEffect`에서 `async/await`를 쓰지 않고 `Promise`를 이용해서 API 요청을 보내고 응답을 받게 작성한 이유에 대해서 였습니다.

의존성에 대한 이야기가 아니었어요! 😄

Reply:
그런데 말씀하신 내용을 보고 의존성을 보니 문제가 있어보이네요.

최초 `userRentals`, `userOverdueRentals`는 데이터가 비어있을거고, 말씀하신 것 처럼 최초에 한번 랜더링 될 때 useEffect가 돌아서, api 요청을 보낼거에요.

그런데, `fetchGroupMe`의 성공 응답에서 `userRentals`, `userOverdueRentals`가 변경되고, 그러면 의존성에 따라서 useEffect가 한번 더 발생하면, 서버로 요청을 한번 더 보낼 것 같아요.

테스트해보시면 서버로 요청이 2번가지 않던가요?

Reply:
![image](https://user-images.githubusercontent.com/29908256/94154714-26b4a800-feb9-11ea-9cca-99cf3d57ab87.png)

생각보다 요청이 더 많이 발생하는것 같습니다....async await해야할것 같긴 한데, 현재 api에서 특정 데이터값을 뽑아오기만 하는 함수를 또다시 function api로 만드는게 효율적으로 안느껴지는것 같습니다ㅠ 
제가 이해한 모스의 제안은 fetchUserProfile의 response값중 name만 필요로 하다면 api자체를 return name 만 하게 만들라 하는 의도이신것 같은데, 다른 뷰에서는 name, userId 식으로 필요하다면 또다른 function을 만들어야 하는걸까요?

Reply:
일단, 주제가 3가지가 섞였습니다.

1. Promise보다 async/await를 쓰자
2. API를 호출하는 함수에 대해서 1)호출하는 함수 2)성공 실패를 처리하는 함수 2개를 만들자.
3. `useEffect`의 dependency가 이상하다.

일단 질문 주신거는 2번에 대해서인데, 이 부분은 일반적인 케이스를 말씀드린 것이며, 효율적으로 느끼지 않는다면 반영하지 않으셔도 상관없습니다. 조금 더 복잡한 로직을 개발한다면 필요로 할 것이라 일단은 넘어가시면 될 것 같습니다.

여기서 1번에 대해서는 로직과는 상관없이 컨벤션과 관련된 이슈입니다. [여기 코멘트](https://github.com/our-library/library-web/pull/4#discussion_r493702193)에서 Promise보다 async/await를 사용하는게 좋을 것 같다고 이유를 말씀드렸고 확인하시고 의견 주세요.

문제가 되는건 3번입니다.
API요청을 `useEffect`를 통해서 하고있고, API를 호출해서 데이터를 업데이트하고있고, 업데이트 되는 데이터는 `useEffect`의 의존성으로 주입되어있습니다.
이러면, useEffect 동작 -> API 호출 ->  데이터 업데이트 -> 데이터가 변경되었으므로 useEffect 동작 -> API호출 -> 데이터 업데이트 ... 이런 식으로 무한루프를 돌게 될 수 있습니다. 
따라서 이 부분은 해결을 하고 넘어가야합니다.

Reply:
[말씀해주신 코멘트](https://github.com/our-library/library-web/pull/4#discussion_r493702193) 를 반영해보았는데 이런 개념이 맞는지 체크해주시면 감사하겠습니다..!

![image](https://user-images.githubusercontent.com/29908256/94164068-8c0d9680-fec3-11ea-94b8-ae2dbc5fe502.png)

![image](https://user-images.githubusercontent.com/29908256/94164286-d262f580-fec3-11ea-9063-b70eee791868.png)